### ISSUE TITLE
Preserve existing layers when switching, or add beforeChange and onChange callback options

### ISSUE THREAD

**Issue by chriszrc** [el/style-switcher#9]
I found it surprising that when changing the map style, all my existing layers were blown away (I'm suprised mapbox can't do this automatically). The best I can tell, this appears to currently be the best way to preserve layers on the mapbox map when changing styles:

https://github.com/mapbox/mapbox-gl-js/issues/4006#issuecomment-368273916

It would be amazing if this could be implemented generically, or as a fallback, at least add change hooks so that users can cache the existing layers before a style change, and then reapply them after the style change has been made-

**Comment by dsgh** [comment]
I had the same issue, was able to workaround it for now by doing:

```js
map.on('styledata', () => {
  if (map.getSource('mySource')) return
  map.addSource('mySource', { ...})
  map.addLayer({ source: 'mySource', ...})
})
```

**Comment by chriszrc** [comment]
@dsgh Yes, re-adding the source and layer is a workaround, but it can be very painful when you've added multiple layers, and when layers are added non-deterministically. In that case, the only source of truth is the current state of what's already been aded. The above link has two suggestions, either cache the sources/layer before the style change and reapply, or try to have the satellite layer already added, and just show hide that layer. I'd like to look more into that option-

**Comment by el** [comment]
Although I agree this is a real issue that should be addressed, I don't believe this should be a part of this package. But instead should be solved in the mapbox-gl package.

**Comment by chriszrc** [comment]
@el I whole-heartedly disagree, I can't really even think of a circumstance when I've added layers to the map, and I want them to disappear just because I changed the base layer. Mapbox is a real outlier here, google maps, leaflet, etc., automatically preserve the layers when changing the base map. 

I've created a fork with a new option for "preserveLayers":

https://github.com/chriszrc/style-switcher

It automatically retains all user-added layers/sources while switching between road/satellite/etc styles. I think I've covered all the edge cases, but if you try it out and find any problems let me know and then maybe I will create a pull request-
